library/std/src/sys/unix/args.rs

```diff
@@ -142,9 +142,20 @@ mod imp {
             let argv = ARGV.load(Ordering::Relaxed);
             let argc = if argv.is_null() { 0 } else { ARGC.load(Ordering::Relaxed) };
             (0..argc)
-                .map(|i| {
-                    let cstr = CStr::from_ptr(*argv.offset(i) as *const libc::c_char);
-                    OsStringExt::from_vec(cstr.to_bytes().to_vec())
+                .filter_map(|i| {
+                    let ptr = *argv.offset(i) as *const libc::c_char;
+                    // Some C commandline parsers are replacing already handled arguments in `argv`
+                    // with `NULL` so it's necessary to skip over them here.
+                    //
+                    // We only worry about that on glibc as that is the only situation where Rust
+                    // is not in charge of `main()` but can still retrieve `argc`/`argv` via
+                    // `.init_array`.
+                    if cfg!(all(target_os = "linux", target_env = "gnu")) && ptr.is_null() {
```

Comment:
On other libc's someone may use a C commandline parser from their rust main function, right? In that case argv would still be modified after libstd stored argc/argv during it's pre-main initialization.

Reply:
You would have to actually get access to the `argv` pointer (it's not passed into `main()`), and then pass it to a C function (via FFI, so `unsafe`).

I'm fine with doing the `NULL` check everywhere and always if we want to allow the above. That `NULL` check is not really expensive and commandline argument parsing is probably not a very performance critical aspect of any application where one additional `NULL` check per argument would hurt :)

Reply:
On macOS there is `_NSGetArgv()`. In fact libstd uses this function.

Reply:
We allocate each argument anyways, so this tiny check shouldn't matter. It's safer to just always do it.

Reply:
Ok then :)